src/codegen/compile.ml

```diff
@@ -1602,12 +1608,30 @@ module Tagged = struct
     assert (!Flags.gc_strategy = Flags.Incremental);
     1l
 
-  (* Note: Post allocation barrier must be applied after initialization *)
+  (* Note: post-allocation barrier must be applied after initialization *)
   let alloc env size tag =
+    assert (size > 1l);
     let name = Printf.sprintf "alloc_size<%d>_tag<%d>" (Int32.to_int size) (Int32.to_int (int_of_tag tag)) in
+    let overflow_mask n =
+      let n = Int32.to_int n in
+      Int32.(logand 0xFFFFl (shift_left minus_one (16 - Numerics.Nat16.(to_int (clz (of_int n)))))) in
+
     Func.share_code0 env name [I32Type] (fun env ->
-      let (set_object, get_object) = new_local env "new_object" in
-      Heap.alloc env size ^^
+      let set_object, get_object = new_local env "new_object" in
+      let size_in_bytes = Int32.(mul size Heap.word_size) in
+      (if !Flags.gc_strategy <> Flags.Incremental && size_in_bytes < 0x8000l then
+         GC.get_heap_pointer env ^^
+         compile_sub_const 1l ^^ (* skew *)
+         GC.get_heap_pointer env ^^
+         compile_add_const size_in_bytes ^^
+         GC.set_heap_pointer env ^^
+         GC.get_heap_pointer env ^^
+         compile_bitand_const (overflow_mask size_in_bytes) ^^
+         G.if0
+           G.nop (* no page crossing *)
+           (Heap.alloc env 0l ^^ G.i Drop) (* ensure that the page is allocated *)
```

Comment:
Would it be possible to economize this `0` allocation by always having the current page allocated?

Reply:
Maybe, but is there a good/better place to do that?

Reply:
I am not sure. Could we allocate the first page in the dynamic heap (in the case the start address would be page-aligned) and then allocate whenever a page is crossed? But maybe it is not worth the complexity...

Reply:
Ok, sorry for the confusion, now I see that the case is only happening when page is crossed (with some false positives).